### ISSUE TITLE
Tenant used as key problem

### ISSUE THREAD

**Issue by dazinator** [saaskit/saaskit#76]
The `TTenant` is being used as the key to the concurrent dictionary here: https://github.com/saaskit/saaskit/blob/master/src/SaasKit.Multitenancy/Internal/TenantPipelineMiddleware.cs#L15

However, when the `TenantContext` is disposed (memory cache invalidation) the `TTenant` instance is disposed. However it's still left as the key in the concurrent dictionary..  Unclear if that will cause problems, but I am raising it as it seems undesirable?

Also there aren't any constraints that I can see, that make it obvious to implementers of TTenant that they need to support being used a dictionary key. (implement GetHashCode, Equals etc))

**Comment by dazinator** [comment]
The more I think about this, if you are in the area of breaking changes, perhaps you would consider forcing TTenant to implement some sort of "Key" or "Id" property, which can be used to identify a tenant.

This could be used as a key for the Tenant's Container cache, and the Middleware Pipeline cache, rather than the TTenant itself.

**Comment by benfoster** [comment]
I didn't want to impose any requirements on the tenant type. This is why
the caching middleware takes a delegate to obtain the tenant identifier
meaning no need to force an ID property.

**Comment by dazinator** [comment]
Yep that makes sense.. but then TTenant is used directly as the key in the line shown without a chance for a key to be provided like you can with the caching implementation.. hasnt seemed to cause me a problem but now I know about it through looking at the code I thought its best to implement Equals and gethashcode just in case.

**Comment by mgaspar** [comment]
I also found out this behavior last week when I was doing some profiling and saw the huge number of instances in the pipeline dictionary. I implemented Equals and GetHashcode to workaround this (worked fine for me, because I was not storing any state in the Tenant itself, but caching state in other services I have more control of). But I would say this behavior is not intuitive, and may lead to issues if people are not aware of this.

**Comment by tillig** [comment]
I ended up using a `string` as my `TTenant` and put all the actual tenant config/settings into properties on `TenantContext`.

**Comment by tillig** [comment]
...but now I'm seeing #68 and wondering if what I'm doing is such a good idea in light of the auto-disposal of properties. Could be a huge gotcha later. Hmmm. May need to fork.
